File: String/443_string_compression.cpp
class Solution {
private:
    int writeGroup(vector<char>& chars, int index, char c, int count) {
        chars[index++] = c;
        if(count > 1) {
            string count_str = to_string(count);
            for(char digit: count_str) {
                chars[index++] = digit;
            }
        }
        return index;
    }
    
public:
    int compress(vector<char>& chars) {
        char prev = chars[0];
        int count = 0;
        int index = 0;

        for(char c: chars) {
            if(c != prev) {
                index = writeGroup(chars, index, prev, count);
                prev = c;
                count = 1;
            } else {
                count += 1;
            }
        }

        index = writeGroup(chars, index, prev, count);
        return index;
    }
};

File: String/443_string_compression.py
class Solution:
    def compress(self, chars: List[str]) -> int:
        prev, count = None, 0
        index = 0

        for c in chars:
            if prev is None:
                prev = c
                count = 1
            elif c == prev:
                count += 1
            else:
                chars[index] = prev
                index += 1
                if count != 1:
                    str_rep = str(count)
                    chars[index: index + len(str_rep)] = list(str_rep)
                    index += len(str_rep)
                prev = c
                count = 1

        chars[index] = prev
        index += 1
        if count != 1:
            str_rep = str(count)
            chars[index: index + len(str_rep)] = list(str_rep)
            index += len(str_rep)

        return index